projects/bp-gallery/src/hooks/mouse-position.hook.ts

```diff
@@ -0,0 +1,48 @@
+import { isEqual } from 'lodash';
+import { useEffect, useState, useCallback } from 'react';
+import { flushSync } from 'react-dom';
+
+type Position = [number, number];
+type MousePosition = {
+  client: Position;
+};
+const listeners = new Set<(position: MousePosition) => void>();
```

Comment:
Wozu braucht ihr das listeners set? Was hindert euch daran direkt den eventlistener im useEffect zu adden und removen?

Reply:
Das war ein Optimierungsgedanke: Ich denke ein EventListener, der alles ausrechnet und an ein Set von listenern gibt ist performanter als viele EventListener, die alle das gleiche machen.

Reply:
> Das war ein Optimierungsgedanke: Ich denke ein EventListener, der alles ausrechnet und an ein Set von listenern gibt ist performanter als viele EventListener, die alle das gleiche machen.

Ihr spart euch letztendlich glaube nur den check nach der Position, der sowieso nur eintritt wenn bei einen mousemove event die mouse nicht moved. In allen anderen Fällen habt ihr genauso viele Aufrufe. Ich glaube die minimale Optimierung ist den doppelt so langen und komplizierteren code nicht wert

Reply:
Ich habs einfach mal getestet:
```js
let clickCount = 0;
for (let i = 0; i < 1_000; i++) {
  document.body.addEventListener("click", () => {
    clickCount++;
  });
}
let start = performance.now();
for (let i = 0; i < 100_000; i++) {
  document.body.click();
}
console.log(performance.now() - start, clickCount);
```
vs.
```js
let clickCount = 0;
let listeners = new Set();
document.body.addEventListener("click", () => {
  for (const listener of listeners) {
    listener();
  }
});
for (let i = 0; i < 1_000; i++) {
  listeners.add(() => {
    clickCount++;
  });
}
let start = performance.now();
for (let i = 0; i < 100_000; i++) {
  document.body.click();
}
console.log(performance.now() - start, clickCount);
```
In Firefox und Chromium hat das zweite immer etwas um 3.000ms ausgegeben.
Das erste dagegen in Firefox um 28.000ms und in Chromium sogar um 130.000ms.
Ich bin gespannt, ob das nur bei mir so ist ;)

Reply:
Okay, scheint wirklich vorallem bei hohen Zahlen einen großen Unterschied zu machen. Krass dass das auch so browser-abhängig ist 🤔